pkg/yurttunnel/dns/node.go

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2021 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"regexp"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/projectinfo"
+	"github.com/openyurtio/openyurt/pkg/yurttunnel/constants"
+)
+
+func (dnsctl *coreDNSRecordController) addNode(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		return
+	}
+	if node.DeletionTimestamp != nil {
+		dnsctl.deleteNode(node)
+		return
+	}
+	klog.V(2).Infof("enqueue node add event for %v", node.Name)
+	dnsctl.enqueue(node, NodeAdd)
+}
+
+func (dnsctl *coreDNSRecordController) updateNode(oldObj, newObj interface{}) {
+	oldNode, ok := oldObj.(*corev1.Node)
+	if !ok {
+		return
+	}
+	newNode, ok := newObj.(*corev1.Node)
+	if !ok {
+		return
+	}
+
+	oldIsEdgeNode, newIsEdgeNode := isEdgeNode(oldNode), isEdgeNode(newNode)
+	if oldIsEdgeNode == newIsEdgeNode {
+		klog.V(2).Infof("skip node update event for %v as node role is not changed", newNode.Name)
+		return
+	}
+
+	klog.V(2).Infof("enqueue node update event for %v, will update dns record", newNode.Name)
+	// shared NodeAdd processing logic
+	dnsctl.enqueue(newNode, NodeAdd)
+}
+
+func (dnsctl *coreDNSRecordController) deleteNode(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("can not get object from tombstone %#v", obj))
+			return
+		}
+		node, ok = tombstone.Obj.(*corev1.Node)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object is not a node %#v", obj))
+			return
+		}
+	}
+	klog.V(2).Infof("enqueue node delete event for %v", node.Name)
+	dnsctl.enqueue(node, NodeDelete)
+}
+
+// onNodeAdd processes node add or update event
+func (dnsctl *coreDNSRecordController) onNodeAdd(node *corev1.Node) error {
+	klog.V(2).Infof("adding node dns record for %v", node.Name)
+	ip, err := getNodeHostIP(node)
+	if err != nil {
+		return err
+	}
+	if isEdgeNode(node) {
+		ip, err = dnsctl.getTunnelServerIP()
```

Comment:
how about get clusterIP of kube-system/x-tunnel-server-svc service only once instead of getting clusterIP per nodeEvent?